### PR TITLE
[RDY] vim-patch:8.1.{265,271,273,274,275,277,278,279,280,281,282,284,286,291,295,296,320,321,339,351,392,399,552}

### PR DESCRIPTION
**Pull request by aufarg**
Hello, this is my first time porting patches. I made some changes from the original diff, notably:
1. Change incsearch state argument name from `is_state` to `s`
2. I'm reusing `command_line_next_incsearch` for `may_adjust_incsearch_highlighting`
3. `may_do_incsearch_highlighting` returns a value

### CODE REVIEW

**Comment by aufarg** [comment]
CI error seems not related to the changes

File: src/nvim/ex_getln.c
Comment (janlazo) [inline src/nvim/ex_getln.c:263]: 2-space indent.

File: src/nvim/ex_getln.c
Comment (janlazo) [inline src/nvim/ex_getln.c:274]: bool

**Comment by janlazo** [comment]
Did you make the changes based on later patches? If so, I suggest to port them over. There are many follow-up patches to the `incsearch` feature.

**Comment by janlazo** [comment]
Ported patch looks good aside from style issues and C99 tweaks (ie. bool).

**Comment by aufarg** [comment]
I've added some later patches. I changed my mind about making `may_do_incsearch_highlighting` returns a value, instead now it runs when `inccommand` highlighting is not being done, since IIUC `inccommand` has its own highlighting.

**Comment by teto** [comment]
Thanks @aufarg for the welcome contribution ! and kudos to @janlazo for helping you throughout, could you "Approve" the PR (via the github ui) if you think it's ready ? I will then try to run it on my machine.

**Comment by janlazo** [comment]
@aufarg Can you add more `incsearch` patches, found by `vim-patch.sh -l` with `--grep incsearch` or `-G insearch`, up to `vim-patch:8.1.0399`? It's fine if you port only the remaining 8.1.02xx patches. I can pick up the rest. Merging your PR as is would mean that a contributor or maintainer has to immediately port the rest to avoid unexpected issues.

File: src/nvim/ex_getln.c
Comment (janlazo) [inline src/nvim/ex_getln.c:2151]: Does `incsearch` interfere with `inccomand`?

File: src/nvim/testdir/test_search.vim
Comment (janlazo) [inline src/nvim/testdir/test_search.vim:389]: Missing tests from patch 8.1.0274.

**Comment by janlazo** [comment]
Everything else looks fine based only on the code.

File: src/nvim/ex_getln.c
Comment (aufarg) [reply src/nvim/ex_getln.c:2151]: They uses different kind of highlighting and kind of interfere with each other. I tried by doing this at first:
```
diff --git a/src/nvim/ex_getln.c b/src/nvim/ex_getln.c
index c58c9b172..bec12ab3b 100644
--- a/src/nvim/ex_getln.c
+++ b/src/nvim/ex_getln.c
@@ -2116,6 +2116,7 @@ static int command_line_changed(CommandLineState *s)
     }
   }
 
+  may_do_incsearch_highlighting(s->firstc, s->count, &s->is_state);
   // 'incsearch' highlighting.
   if (s->firstc == ':'
       && current_sctx.sc_sid == 0    // only if interactive
@@ -2147,8 +2148,6 @@ static int command_line_changed(CommandLineState *s)
     State = (State & ~CMDPREVIEW);
     close_preview_windows();
     update_screen(SOME_VALID);  // Clear 'inccommand' preview.
-  } else {
-    may_do_incsearch_highlighting(s->firstc, s->count, &s->is_state);
   }
 
   if (cmdmsg_rl || (p_arshape && !p_tbidi && enc_utf8)) {
```
If both run at the same time, `may_do_incsearch_highlighting` make the first match's `reverse` attribute set to true, and it fails some of the existing tests. I also tried with having a return value returned by `may_do_incsearch_highlighting`, so that only when it fails we do the `inccommand` block without success.

It also fails some other tests like `:substitute, inccommand=split deactivates if 'redrawtime' is exceeded #5602`, but I can't figure out why.

I figure that since before patch 8.1.0271 `incsearch` doesn't work with `:s`, that means whenever `inccommand` was set, we never uses the highlighting from `incsearch` on `:s` and instead we use the highlighting from `inccommand` block. So I thought we can keep current behaviour by not using `incsearch` highlighting on `:s` and use `inccommand` highlighting like we did before.

File: src/nvim/testdir/test_search.vim
Comment (aufarg) [reply src/nvim/testdir/test_search.vim:389]: For this hunk:
```
diff --git a/src/nvim/testdir/test_search.vim b/src/nvim/testdir/test_search.vim
index 71de172dc..7e8ddbe64 100644
--- a/src/nvim/testdir/test_search.vim
+++ b/src/nvim/testdir/test_search.vim
@@ -516,11 +522,11 @@ func Test_search_cmdline7()
   endif
   " need to disable char_avail,
   " so that expansion of commandline works
   call test_override("char_avail", 1)
   new
-  let @/='b'
+  let @/ = 'b'
   call setline(1, [' bbvimb', ''])
   set incsearch
   " first match
   norm! gg0
   " moves to next match of previous search pattern, just like /<cr>
/
```
It changes the line on `Test_search_cmdline7`, which was not ported from mainline. It was introduced in patch 8.0.1250. Should I port the whole test? Should it be part of this changeset or should I make another PR?

**Comment by aufarg** [comment]
> Vim doesn't accept `^T` delimiter since patch 8.1.0271 so `:%s^Tfoo^Tbar^Tg` does not work. Does that break any use case?

I'm not really sure because I just found that Vim does accept other delimiter other than `/` (The help on `:s` doesn't mention anything about the delimiter). How do I go about finding out if it break any use case?

File: src/nvim/testdir/test_search.vim
Comment (janlazo) [reply src/nvim/testdir/test_search.vim:389]: Port the whole test. It was omitted because of `test_override` but that makes it harder to port patches.

**Comment by janlazo** [comment]
@aufarg 

> @janlazo
> 
> > Can you add more incsearch patches, found by vim-patch.sh -l with --grep incsearch or -G insearch, up to vim-patch:8.1.0399? It's fine if you port only the remaining 8.1.02xx patches. I can pick up the rest. Merging your PR as is would mean that a contributor or maintainer has to immediately port the rest to avoid unexpected issues.
> 
> Sure, I'm fine with that, but I do have question regarding patches unrelated to `incsearch` that also needs to be ported. In fact, I stopped 8.1.0280 because 8.1.0282 introduce patch to `ex_docmd.c` which seems to miss some of previous patches. Should I also port them or not?

Can you list those missing patches?

> > Vim doesn't accept `^T` delimiter since patch 8.1.0271 so `:%s^Tfoo^Tbar^Tg` does not work. Does that break any use case?
> 
> I'm not really sure because I just found that Vim does accept other delimiter other than `/` (The help on `:s` doesn't mention anything about the delimiter). How do I go about finding out if it break any use case?

That question was meant for Neovim core team and whoever contributed to `inccommand` feature. /cc @KillTheMule 
Check Vim's github issues about incsearch or delimiter. If you find nothing, then it's probably fine.

**Comment by KillTheMule** [comment]
> That question was meant for Neovim core team and whoever contributed to `inccommand` feature. /cc @KillTheMule

I don't have any special insight here, there's nothing in inccomand that would make ^T of special use. I guess if vim does it we should be doing it, too.

**Comment by aufarg** [comment]
@janlazo 
>
> Can you list those missing patches?
> 

These are the missing patches for `ex_docmd.c` before v8.1.0282:
```
  • v8.0.1459: cannot handle change of directory
  • v8.0.1505: debugger can't break on a condition
  • v8.0.1570: can't use :popup for a menu in the terminal
  • v8.0.1592: terminal windows in a session are not properly restored
  • v8.0.1804: using :normal in terminal window causes problems
  • v8.1.0035: not easy to switch between prompt buffer and other windows
  • v8.1.0095: dialog for ":browse tabnew" says "new window"
  • v8.1.0149: session is wrong with multiple tabs when :lcd was used
  • v8.1.0228: dropping files is ignored while Vim is busy
  • v8.1.0281: parsing command modifiers is not separated
```

**Comment by janlazo** [comment]
Which patch caused the test failures?

**Comment by aufarg** [comment]
> Which patch caused the test failures?

All of the failures seems to be caused by 8.1.0296.

**Comment by aufarg** [comment]
I added one patch to merge `parsed_state_T` on `ex_docmd.c` to `exarg_T`. Will merge with 8.1.0281 if the patch is okay.

**Comment by janlazo** [comment]
If you break this PR to 2 branches, one for patches before and one for patches after patch 8.1.0296 (or patch 8.1.0281 because of your patch), are there patches in the latter branch than you can cherry-pick to the former branch with minimal or no difficulty and without failing the tests? 

I can't merge the PR if there are failing tests.

**Comment by justinmk** [comment]
re https://github.com/neovim/neovim/pull/12721#discussion_r471073908 

> I figure that since before patch 8.1.0271 `incsearch` doesn't work with `:s`, that means whenever `inccommand` was set, we never uses the highlighting from `incsearch` on `:s` and instead we use the highlighting from `inccommand` block. So I thought we can keep current behaviour by not using `incsearch` highlighting on `:s` and use `inccommand` highlighting like we did before.

Yes, that is the right approach.

**Comment by aufarg** [comment]
When I try to test with UBSan in local, it always failed with:
```
[----------] Running tests from test/functional/legacy/search_spec.lua
-- Tests exited non-zero: SIGPIPE
-- No output to stderr.
```
Any pointers?

**Comment by aufarg** [comment]
Found the fix for the leak detected by UBSan in v8.1.0552. I've added it to the list of the patches.

File: src/nvim/ex_getln.c
Comment (janlazo) [inline src/nvim/ex_getln.c:159]: bool

File: src/nvim/ex_getln.c
Comment (janlazo) [inline src/nvim/ex_getln.c:158]: bool

File: src/nvim/globals.h
Comment (janlazo) [inline src/nvim/globals.h:356]: in general, if the variable is set to true/false only, use bool always. If it is set to `MAYBE`,  use TriState. If it's set to anything else, keep int and use integer values directly even if Vim uses true,false for 1,0.

File: src/nvim/ex_docmd.c
Comment (janlazo) [inline src/nvim/ex_docmd.c]: Is this removed because `parse_one_cmd` doesn't do as documented or is it useless or is it to make patching easier?

**Comment by janlazo** [comment]
Tests added/updated in test_search.vim are not in sync with Vim.
- Test_search_cmdline3s

File: src/nvim/ex_docmd.c
Comment (janlazo) [reply src/nvim/ex_docmd.c]: Did you check https://github.com/neovim/neovim/pull/11340?

File: src/nvim/testdir/test_search.vim
Comment (janlazo) [inline src/nvim/testdir/test_search.vim:715]: In order to verify that your changes work, convert these screendump tests into lua functional screen tests in `test/functional/`. These codecov warnings appear because these screendump tests are skipped.

You can ignore the `test_override` tests. As far as I know, Neovim does not support modifying the internal state, at least within Vimscript.

File: src/nvim/ex_docmd.c
Comment (aufarg) [reply src/nvim/ex_docmd.c]: I removed `parse_one_cmd` because it seems that the function does the same thing as the additional code from v8.1.0281 on `do_one_cmd`, except it brings no side-effect to the globals. The function is called only in one place though, so I wonder if making the function have no side-effect warrants the additional effort to port later patches. It also seems that the patch achieve the same goal as #11340.

That said, I'm not aware if the team have something in mind that requires the function to not do any side-effect. Let me know if I shouldn't remove it.

File: src/nvim/testdir/test_search.vim
Comment (aufarg) [reply src/nvim/testdir/test_search.vim:721]: I've added the screendump tests, not sure if I put it in the correct place though. Also the codecov warnings are still there.

File: src/nvim/ex_getln.c
Comment (janlazo) [inline src/nvim/ex_getln.c:275]: true

File: src/nvim/testdir/test_search.vim
Comment (janlazo) [inline src/nvim/testdir/test_search.vim:512]: Duplicate `+incsearch` check?

**Comment by aufarg** [comment]
I've changed all occurence of `TRUE` on my added lines to `true` and changed `delim_optional` to be `bool`. I think that's all that I can find that should be changed from the original patch.

**Comment by janlazo** [comment]
I'll review again and approve if it's ready. Good work!

**Review by janlazo — APPROVED** [review]
LGTM

**Comment by janlazo** [comment]
@teto Are you testing this PR?

**Comment by teto** [comment]
@janlazo no need if you approved of it ;) I'll let you merge since you have the rights now :)

File: src/nvim/ex_docmd.c
Comment (janlazo) [reply src/nvim/ex_docmd.c]: Will have to revist this for recent Vim 8.2 patches such as https://github.com/vim/vim/commit/5661ed6c833e05467cab33cb9b1c535e7e5cc570.

**Pull request merged**
